### DISCUSSION TITLE
[Done] Apple Silicon (M1) supported 🍎

### DISCUSSION THREAD

**Discussion by JanMP** [meteor/meteor#11633] [Feature Requests]
Trying to install meteor on new MacBook Pro sadly yields this result:
```
➜ curl https://install.meteor.com/ | sh

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7915    0  7915    0     0  14390      0 --:--:-- --:--:-- --:--:-- 14364
Only 64-bit Intel processors are supported at this time.
```

**Comment by wreiske** [comment]
<img width="710" alt="Screen Shot 2020-11-20 at 9 30 20 AM" src="https://user-images.githubusercontent.com/6295044/99830764-06007b80-2b13-11eb-8ef3-cfce546d9d8e.png">

Seeing the same thing.

**EDIT:**

Try this:

```
arch -x86_64 /bin/sh -c "$(curl -fsSL https://install.meteor.com/)"
```

You can also try opening up a terminal, type the following, and then run everything as you normally would.
```
arch -x86_64 zsh
curl https://install.meteor.com/ | sh
```

<img width="1390" alt="Screen Shot 2020-11-20 at 9 40 02 AM" src="https://user-images.githubusercontent.com/6295044/99831744-693edd80-2b14-11eb-95f6-45795513db53.png">

**Comment by hexsprite** [comment]
Until there's a native Apple Silicon version, will running it in a Rosetta terminal work? https://osxdaily.com/2020/11/18/how-run-homebrew-x86-terminal-apple-silicon-mac/

**Comment by JanMP** [comment]
Thanks guys, both your solutions work like a charm. And despite running in Rosetta it still is quite a bit faster than my 2016 MacBook Pro 13'' 🎉

**Comment by renanccastro** [comment]
We still don't have native binaries for nodejs on Apple silicon's. It's important to keep an eye here: https://github.com/nodejs/build/issues/2474

I think a good workaround we can do right now is to change the meteor script to use -arch (rosetta 2) when it detects the arm architecture.

**Comment by lemontron** [comment]
Homebrew is now hosting a native arm64 binary of nodejs.

https://homebrew.bintray.com/bottles/node-15.6.0.arm64_big_sur.bottle.tar.gz

**Comment by zvictor** [comment]
> Homebrew is now hosting a native arm64 binary of nodejs.
> https://homebrew.bintray.com/bottles/node-15.6.0.arm64_big_sur.bottle.tar.gz

Would it then be possible to run meteor using a different node binary than the bundled one?
Can I make Meteor use this node binary?

**Comment by lemontron** [comment]
I dove into the `METEOR DEV BUNDLE` and tried to change it to grab node from the bottle. My thoughts were that Meteor has always bundled node since the beginning, so it's kind of unlikely that meteor would offer a feature to run the default node binary, so the most fitting way would be to get it from the bottle. But it was just outside of my area of expertise. I might have another go at it.

**Comment by satyavh** [comment]
Even though this allows you to install Meteor, it still means Meteor needs to run in Rosetta mode, losing performance.

**Comment by StorytellerCZ** [comment]
Looks like the support on this is going to be for Node 15+, so if we rush it, then we could get it to Meteor with Node 16 probably in July, but that is rushing it a lot.

**Comment by gainmedia-sakib** [comment]
### **it still not working for me** 

❯ arch -x86_64 /bin/sh -c "$(curl -fsSL https://install.meteor.com/)"
Only 64-bit Intel processors are supported at this time.

❯ arch -x86_64 zsh
curl https://install.meteor.com/ | sh

Only 64-bit Intel processors are supported at this time.

### **Already I trying some method but it's not working, trying Rosetta Terminal i get same error **
### anyone sagest me how can I do this?

**Reply by filipenevola** [reply]
>  anyone sagest me how can I do this?

How are you running with Rosetta? It's working fine with Rosetta.

**Comment by StorytellerCZ** [comment]
Since the support for Apple Silicon is from Node 16 we will be able to tackle this once we update there. Given that there are some major hurdles to get over in order to get Meteor to Node 16 don't expect this to happen till the end of the year at earliest (by my estimation).

**Comment by lemontron** [comment]
Is there a branch for the Node 16 project? I just want to play with it, I'll happily publish whatever I make.

**Comment by StorytellerCZ** [comment]
Node 16 presents breaking changes for Meteor, but @filipenevola is working on them. We'll probably start one after 2.3 release, but don't expect it to work for a while, especially on Apple Silicon as I think that will require some extra changes in the CI to account for it.

**Comment by ANUBHAV04091998** [comment]
thank you it worked.!!

**Comment by filipenevola** [comment]
It seems that Node.js 14 is compatible and [has all the necessary code to build for M1](https://github.com/nodejs/build/issues/2474#issuecomment-839683917) but the Node.js team is not building for this architecture.

We could try to find a reliable company building for M1 or build ourselves.

**Comment by lemontron** [comment]
Using brew I was able to download [Node v14.](https://formulae.brew.sh/formula/node@14
) This is great news as it means someone at brew is already building v14. 

Brew creates a tar.gz archive containing the prebuilt binaries ("bottles"), in the folder `~/Library/Caches/Homebrew/downloads/`


👉 [Download Node v14 for Apple Silicon](https://github.com/meteor/meteor/files/7188872/fcbc93fb6e7715ace05a94cd2bf1b2dc6efbd1ec51108a49dc928f831b1c2402--node%4014--14.17.6.arm64_big_sur.bottle.tar.gz)

**Comment by perbergland** [comment]
Some enlightening links:

https://github.com/nodejs/node/pull/38051
https://github.com/nodejs/node/issues/40126

So the work to backport changes has been done and it is in "experimental" mode which is good enough for now, I guess.
This will only be used for development and testing - no one runs anything in production on macOS/ARM right?

**Comment by filipenevola** [comment]
Hi, @renanccastro and I were analyzing the options here and as even nvm is already installing Node.js 14 in M1 we believe we should do the same.

We are working in our CI infrastructure to configure a new M1 machine, so we can start compiling Node.js for M1 as well and start experimenting with it.

Of course, we can find problems along the way so we can't promise any dates for now.

We will update this issue as soon as we have more news.

**Comment by renanccastro** [comment]
Hello!
We already have the Node 14 build pipeline ready in our CI, and the first preliminary builds were successful.
I will be testing generating a new Meteor dev bundle for mac m1. I haven't seen yet Mongo binaries for darwin arm64, but if they do exist we are almost done.

**Comment by renanccastro** [comment]
**Updating:**
Node is built and integrated into our release process, but there are still no binaries for MongoDB that are compatible with darwin arm64.

https://jira.mongodb.org/browse/SERVER-50115

We have to wait on that, or decide that we won't ship embedded mongo for m1. We could see if we can use rosetta only for mongo, when running it from meteor-tool. What do you think?

**Comment by lemontron** [comment]
Even if only Node was optimized, that would be a huge game changer. Currently mongo is unoptimized and seeing a speedup only to Node would be very welcome. Furthermore, embedded mongo does not typically serve high capacities, so there is less of a visible speed impact. Even beyond that, some of us myself included connect to a cloud based or on premise mongo database for development purposes.

**Comment by filipenevola** [comment]
Hey all, good news.

The next version of Meteor is going to support M1 😉 

More details [here](https://github.com/meteor/meteor/pull/11711).

**Comment by lemontron** [comment]
I have it running on my new M1 MacBook Pro. Thanks filipenevola and the team for your hard work!

## Quick How To Guide

Clone meteor devel branch from GitHub
``` bash
cd ~
git clone --recursive https://github.com/meteor/meteor.git
cd meteor
./meteor --help
```

Create an alias to the newly downloaded meteor. On a fresh MacBook Pro with Monterrey, the .zshrc file is now called .zprofile, thanks Apple...
``` bash
nano ~/.zprofile
alias meteor=$HOME/meteor/meteor
```

Restart terminal and you are good to go. WebStorm seems to get stuck on `Running Meteor from a checkout -- overrides project version (Meteor 2.5)` (probably not a new occurrence) but launching from Terminal works.

**Reply by filipenevola** [reply]
thank you for the feedback.

**Comment by rglover** [comment]
May be help to some people here: @wreiske's instructions work great if you open your terminal in Rosetta (set this via the "Get Info" window of the app you want to run in Rosetta—see: https://www.byran.tech/html/how-to-make-a-rosetta-2-emulated-x86-terminal-on-arm-apple-silicon-chips.html). Thanks @filipenevola for the tip.

**Comment by renanccastro** [comment]
Hi, everyone! Meteor 2.5.1-beta.1 is out with M1 support.

For installing it, make sure "meteor" is uninstalled from your NPM with:

`meteor-installer uninstall && npm uninstall -g meteor`

Then, install it with:

**`npm install -g meteor@2.5.1-beta.1`**

Please provide any feedback you have in the 2.5.1 PR, here: https://github.com/meteor/meteor/pull/11748

**Comment by renanccastro** [comment]
Updates: We've identified that the implementation of the fiber built by the core package node-fibers was using the CORO_PTHREAD backend in arm and arm64 environments, and this backend is slow and is not working well (it's failing tests and have some dead-locks issue).

From my debugging, it seems that when a Fiber is marked for deletion, the backing thread is being used in a context transfer operation resulting in a pthread_cond_init in an uninitialized variable. I have not proceeded to fix it, as I was able to use the ucontext backend.

I was able to compile fibers using CORO_UCONTEXT and pass a special directive to the node-gyp compiling configuration, and this implementation is working well it seems.

It would be really great if we could have more people testing it on M1.

Let me know if you find any more issues on 2.5.1-beta.3.

For installing, please uninstall all other meteor versions first:

```
meteor-installer uninstall (or remove ~/.meteor)
npm uninstall -g meteor

```
then install the beta 3 with:

`
npm install -g meteor@2.5.1-beta.3
`

**Comment by filipenevola** [comment]
I created a new app in my M1 machine and also in my Intel machine, see machine specs below (my Intel is much better except the processor, of course).

Intel Machine

![About This Mac 2021-11-08 13-52-21](https://user-images.githubusercontent.com/437372/140800627-e62acb0c-89bc-4060-a066-a66286a9eebb.png)

M1 Machine

<img width="241" alt="Screen Shot 2021-11-08 at 1 52 51 PM" src="https://user-images.githubusercontent.com/437372/140800692-a92be819-0225-431f-812b-8cc658cb8dbb.png">

Create command:
```
meteor create m1app
```

So first I started my app for the first time, made a few changes (to warm up caches).

Then I started it again with METEOR_PROFILE activated (`METEOR_PROFILE=1 meteor`).

```
m1:    second start 249 ms (Server startup)
intel: second start 466 ms (Server startup)

m1:    server change 249 ms (Server startup)
intel: server change 463 ms (Server startup)

m1:    server change delayed web.browser.legacy in 60ms
intel: server change delayed web.browser.legacy in 118ms

m1:    client change 71 ms (Rebuild app)
intel: client change 372 ms (Rebuild app)
```

As you can see my M1 machine is at least 47% faster.

**Reply by wreiske** [reply]
Yup! Which is why I've deployed a few M1 mac mini servers in a cluster for my meteor apps :) 

I don't use AWS personally, but it would be interesting to benchmark this using something like https://aws.amazon.com/ec2/graviton/

**Comment by filipenevola** [comment]
Expanding to compare Intel, M1 with Rosetta and M1 without Rosetta.

Specs: 
Intel: 2.4 GHz 8-Core Intel Core i9 / 64 GB 2667 MHz DDR4
M1: Apple M1 / 16 GB

Table:
| Action | Processor / Arch | Time (ms) |
|-|-|-|
| second start (Server startup) | intel | 466 |
| second start (Server startup) | m1 rosetta | 398 |
| second start (Server startup) | m1 | 249 |
| | | |
| server change (Server startup) | intel | 463 |
| server change (Server startup) | m1 rosetta | 393 |
| server change (Server startup) | m1 | 249 |
| | | |
| server change delayed web.browser.legacy | intel | 118 |
| server change delayed web.browser.legacy | m1 rosetta | 96 |
| server change delayed web.browser.legacy | m1 | 60 |
| | | |
| client change (Rebuild app) | intel | 372 |
| client change (Rebuild app) | m1 rosetta | 175 |
| client change (Rebuild app) | m1 | 71 |

**Comment by filipenevola** [comment]
Hi, this is solved now.

You can read our blog [post](https://blog.meteor.com/meteor-2-5-1-with-native-mac-m1-support-fb1c4b6990b5) about it.
